ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,328 @@
+        if ident == member {
+            match iterable {
+                Iterable::Identifier(ident) => {
+                    let ident_type = symbol_table.get_type(ident.name())?;
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                i,
+                                CURRENT_ROW,
+                            )));
+                        }
+                        IdentifierType::IntegrityVariable(var_type) => {
+                            match var_type.value() {
+                                VariableType::Vector(vector) => {
+                                    validate_access(i, vector.len())?;
+                                    return Ok(vector[i].clone());
+                                }
+                                // TODO: Handle matrix access
+                                _ => Err(SemanticError::InvalidListComprehension(
+                                    "Iterables should be vectors".to_string(),
+                                ))?,
+                            }
+                        }
+                        IdentifierType::PublicInput(size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                i,
+                            )));
+                        }
+                        IdentifierType::RandomValuesBinding(_, size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                i,
+                            )));
+                        }
+                        _ => Err(SemanticError::InvalidListComprehension(
+                            "Invalid type for a vector".to_string(),
+                        ))?,
+                    }
+                }
+                Iterable::Range(range) => {
+                    return Ok(Expression::Const((range.start() + i) as u64));
+                }
+                Iterable::Slice(ident, range) => {
+                    let ident_type = symbol_table.get_type(ident.name())?;
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                range.start() + i,
+                                CURRENT_ROW,
+                            )));
```

Comment:
Why do we hard-code `CURRENT_ROW` here? Shouldn't the offset come from `trace_columns`?

Reply:
The AST has current row trace columns stored as general element since there's no way to disambiguate between whether it's a trace column or any other symbol table like an intermediate variable. So, if it's stored as a trace column in symbol table, we know that it should have 0 row offset here.

Reply:
I might be missing something here, but how is something like this stored in the symbol table:
```
trace_columns:
    main: [a, b]

integrity_constraints:
    let x = a'
    enf x = b + 2
```

Reply:
The trace columns are stored as:
```
TraceColumns(TraceColumns)
```
where
```
pub struct TraceColumns {
    trace_segment: TraceSegment,
    offset: usize,
    size: usize,
}
```
where offset is the column offset of the trace column group instead of row offset.

When we insert the intermediate variable `let x = a'`, `a'` would be stored in the symbol table as:
```
TraceElement(IndexedTraceAccess { trace_segment: 0, col_idx: 0, row_offset: 1 })
```

I think the confusion might be that `a` and `a'` are stored differently in the AST. When we encounter `a` at the parser level we don't know if it's a trace column or any other identifier so we store it as `Expression::Elem`, however when we encounter `a'` at the parser level, we know that it should be referring it to a trace column so we store it as `Expression::NamedTraceAccess(NamedTraceAccess)` where
```
pub struct NamedTraceAccess {
    name: Identifier,
    idx: usize,
    row_offset: usize,
}
```

Maybe there's a better way to do this since this is certainly a little confusing. Please let me know if you have other suggestion.

So, when we parse the AST expression while unfolding list comprehension, if the expression is of type `Elem`, we know that it should be referring to current row.